python/google/protobuf/_importer_impl.py

```diff
@@ -0,0 +1,216 @@
+from google.protobuf import message as _message
+from google.protobuf.internal import enum_type_wrapper
+from google.protobuf import reflection as _reflection
+from google.protobuf import symbol_database as _symbol_database
+
+_sym_db = _symbol_database.Default()
+
+_PROTO_MODULE_SUFFIX = '_pb2'
+
+def _module_name_to_proto_file(module_name):
+  components = module_name.split('.')
+  proto_name = components[-1][:-1 * len(_PROTO_MODULE_SUFFIX)]
+  return os.path.sep.join(components[:-1] + [proto_name + '.proto'])
+
+def _proto_file_to_module_name(proto_file):
+  components = proto_file.split(os.path.sep)
+  proto_base_name = os.path.splitext(components[-1])[0]
+  return '.'.join(components[:-1] +
+                  [proto_base_name + _PROTO_MODULE_SUFFIX])
+
+@contextlib.contextmanager
+def _augmented_syspath(new_paths):
+  original_sys_path = sys.path
+  if new_paths is not None:
+    sys.path = sys.path + new_paths
+  try:
+    yield
+  finally:
+    sys.path = original_sys_path
+
+class ProtoLoader(importlib.abc.Loader):
+  """Instantiates a module equivalent to a _pb2.py file from a .proto."""
+
+  def __init__(self, module_name, protobuf_path):
+    self._module_name = module_name
+    self._protobuf_path = protobuf_path
+
+  def create_module(self, spec):
+    return None
+
+  def _generated_file_to_module_name(self, filepath):
+    components = filepath.split(os.path.sep)
+    return '.'.join(components[:-1] + [os.path.splitext(components[-1])[0]])
+
+  @staticmethod
+  def _register_message(sym_db, message_descriptor, module_name):
+    """Registers a message in the symbol database."""
+    nested_dict = {
+        '__module__': module_name,
+        'DESCRIPTOR': message_descriptor,
+    }
+    for nested_type_desc in message_descriptor.nested_types:
+      nested_type = ProtoLoader._register_message(sym_db, nested_type_desc,
+                                                  module_name)
+      nested_dict[nested_type_desc.name] = nested_type
+    for nested_enum_desc in message_descriptor.enum_types:
+      sym_db.RegisterEnumDescriptor(nested_enum_desc)
+    message_type = (
+        _reflection.GeneratedProtocolMessageType(message_descriptor.name,
+                                                 (_message.Message,),
+                                                 nested_dict))
+    sym_db.RegisterMessageDescriptor(message_descriptor)
+    sym_db.RegisterMessage(message_type)
+    return message_type
+
+  def exec_module(self, module):
+    """Instantiate a module identical to the generated version."""
+    file_descriptor = _pyext_message.FileDescriptor.FromFile(
+        self._protobuf_path.encode('ascii'),
+        [path.encode('ascii') for path in sys.path])
+    _sym_db.RegisterFileDescriptor(file_descriptor)
+    for dependency in file_descriptor.dependencies:
+      module_name = _proto_file_to_module_name(dependency.name)
+      if module_name not in sys.modules:
+        importlib.import_module(module_name)
```

Comment:
The runtime generation must not affect system libraries. Consider two transitive dependencies, which actually need a different definition for the same (conflicting) name: which one wins?

We cannot say, and we should not try to say. Touching the import machinery is simply too subtle for us to hide behind a function call that can look at arbitrary files.

Reply:
@dlj-NaN Can you expand on the sorts of system libraries you're concerned about? This should only affect "*_pb2" modules. In google3, this module loader will only ever load a module when the user explicitly adds a ".proto" file as a `data` dependency on their `py_binary`. In open source, non-hermetic builds make a conflict more likely, but perhaps also less serious.

> Consider two transitive dependencies, which actually need a different definition for the same (conflicting) name: which one wins?

Can you give an example of such a situation? While I can see this happening for an arbitrary module loader, I'm having a hard time coming up with a situation where this particular module loader would exhibit this behavior. You've have to have a `foo.proto` as well as a `foo_pb2` on your `sys.path`. Then you would need a situation where "(two transitive dependencies)  actually need a different definition for the same (conflicting) name"? I suppose this could happen if you lay down an alternate definition for a ".proto" used in a low-level library? This seems unlikely to happen outside of Google3.

> Touching the import machinery is simply too subtle for us to hide behind a function call that can look at arbitrary files.

Are you against modification of the import machinery *in principle* or just against wrapping it in a function call? Addition of the module loader to `sys.meta_path` could instead be an opt-in process enabled by something like a `protobuf.enable_disk_proto_files()` (bikesheddable name). Note that in the other discussion thread on this PR, @anandolee and I discussed limiting the time that the module loader spends on `sys.meta_path` to the duration of the call to `protobuf.protos()`. How would that affect the situations you're concerned about?

Reply:
@dlj-NaN Friendly ping.